pkg/controller/security/manager.go

```diff
@@ -0,0 +1,265 @@
+ */
+
+package security
+
+import (
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"container/heap"
+
+	istiosecurity "istio.io/istio/pkg/security"
+
+	"kmesh.net/kmesh/pkg/constants"
+	"kmesh.net/kmesh/pkg/logger"
+)
+
+type certExp struct {
+	Identity string
+	exp      time.Time
+	index    int
+}
+
+type certsCache struct {
+	cert   *istiosecurity.SecretItem
+	refCnt int32
+}
+
+type SecretManagerCache struct {
+	caClient *CaClient
+
+	// configOptions includes all configurable params for the cache.
+	configOptions *istiosecurity.Options
+
+	// storing certificates
+	certsCache *sync.Map
+
+	//caRootPath string
+	pending *PriorityQueue
+}
+
+var log = logger.NewLoggerField("security")
+
+type PriorityQueue struct {
+	queue []*certExp
+	mu    sync.Mutex
+}
+
+func NewPriorityQueue() *PriorityQueue {
+	return &PriorityQueue{
+		queue: make([]*certExp, 0),
+		mu:    sync.Mutex{},
+	}
+}
+
+func (pq *PriorityQueue) Push(x interface{}) {
+	n := len(pq.queue)
+	item := x.(*certExp)
+	item.index = n
+	pq.queue = append(pq.queue, item)
+}
+
+func (pq *PriorityQueue) Pop() interface{} {
+	old := pq.queue
+	n := len(old)
+	x := old[n-1]
+	old[n-1] = nil // avoid memory leak
+	x.index = -1
+	pq.queue = old[0 : n-1]
+	return x
+}
+
+func (pq *PriorityQueue) Len() int {
+	return len(pq.queue)
+}
+
+func (pq *PriorityQueue) Less(i, j int) bool {
+	return pq.queue[i].exp.Before(pq.queue[j].exp)
+}
+
+func (pq *PriorityQueue) Swap(i, j int) {
+	pq.queue[i], pq.queue[j] = pq.queue[j], pq.queue[i]
+}
+
+func (pq *PriorityQueue) addItem(certExp *certExp) {
+	pq.mu.Lock()
+	defer pq.mu.Unlock()
+	heap.Push(pq, certExp)
+}
+
+func (pq *PriorityQueue) delete(Identity string) *certExp {
+	pq.mu.Lock()
+	defer pq.mu.Unlock()
+	for i := 0; i < len(pq.queue); i++ {
+		log.Debugf("delete item.Identity:%v\n", pq.queue[i].Identity)
+		if pq.queue[i].Identity == Identity {
+			return heap.Remove(pq, i).(*certExp)
+		}
+	}
+	return nil
+}
+
+func (pq *PriorityQueue) lookTop() *certExp {
+	pq.mu.Lock()
+	defer pq.mu.Unlock()
+	return pq.queue[0]
+}
+
+// NewsecretManager creates a new secretManager.
+func NewsecretManager() (*SecretManagerCache, error) {
+	var certCache sync.Map
+
+	tlsOpts = &TLSOptions{
+		RootCert: constants.RootCertPath,
+	}
+
+	options := NewSecurityOptions()
+	caClient, err := NewCaClient(options, tlsOpts)
+	if err != nil {
+		return nil, err
+	}
+
+	pq := NewPriorityQueue()
+	heap.Init(pq)
+
+	secretManager := SecretManagerCache{
+		caClient:      caClient,
+		configOptions: options,
+		certsCache:    &certCache,
+		pending:       pq,
+	}
+	go secretManager.refreshExpiringCerts()
+	return &secretManager, nil
+}
+
+// Automatically check and refresh when the validity period expires
+// Store the Identity in the priority queue according to the expiration time.
+// Check the highest priority element in the queue every 5 minutes.
+// If it is about to expire, pop up the element and reapply for the certificate.
+func (s *SecretManagerCache) refreshExpiringCerts() {
```

Comment:
1. Because the certificates are SA granularity, there won't be many certificates on Node; add a certificate management thread specifically responsible for the logic of adding, deleting, and modifying certificates, as well as reapplying, to avoid multi-threading issues;
2. When refreshing the workload, notify the certificate management thread through a channel (chan).
```go
go certMng()
	|- case AddCert
	|- case DelCert
	|- case Cert reflush(timeout)
```

Outstanding tasks:
1. When providing the GetCert function in the future, locking operations need to be added to certCache (a custom data structure with locks).
2. Complete the implementation of the Kmesh management API.
```c
workload_is_managed
	|- if (is_mnged(wl_mng_cache))
		|- exist 
			|- return true
		|- notexist
			|- if workload managed in eBPF map
				|- exist
					|- add wl_mng_cache; return true;
				|- notexist
					|- return false;

del_wl_mng_cache
```